### PR TITLE
fix(compiler): optional chaining does not always result in an optional type

### LINKED ISSUE #2965: Optional chaining expression is not assigned an optional type
### I tried this:

```
bring cloud;

let get = inflight(req: cloud.ApiRequest): cloud.ApiResponse => {
  let auth = req.headers?.get("authorization");

  if (auth == nil) {
    log("No authorization header");
  }
  
  return cloud.ApiResponse {
    status: 200,
    headers: {"Content-Type": "application/json"},
    body: Json.stringify({auth: auth})
  };  
};
```

[Playground](https://play.winglang.io/?code=YnJpbmcgY2xvdWQ7CgpsZXQgZ2V0ID0gaW5mbGlnaHQocmVxOiBjbG91ZC5BcGlSZXF1ZXN0KTogY2xvdWQuQXBpUmVzcG9uc2UgPT4gewogIGxldCBhdXRoID0gcmVxLmhlYWRlcnM%2FLmdldCgiYXV0aG9yaXphdGlvbiIpOwoKICBpZiAoYXV0aCA9PSBuaWwpIHsKICAgIGxvZygiTm8gYXV0aG9yaXphdGlvbiBoZWFkZXIiKTsKICB9CiAgCiAgcmV0dXJuIGNsb3VkLkFwaVJlc3BvbnNlIHsKICAgIHN0YXR1czogMjAwLAogICAgaGVhZGVyczogeyJDb250ZW50LVR5cGUiOiAiYXBwbGljYXRpb24vanNvbiJ9LAogICAgYm9keTogSnNvbi5zdHJpbmdpZnkoe2F1dGg6IGF1dGh9KQogIH07ICAKfTs%3D)

### This happened:

```
Failed to compile.

error: Expected type to be "str", but got "nil" instead (hint: to allow "nil" assignment use optional type: "str?")
  --> test.w:6:7
  |
6 |   if (auth == nil) {
  |       ^^^^^^^^^^^ Expected type to be "str", but got "nil" instead (hint: to allow "nil" assignment use optional type: "str?")

```

### I expected this:

No error, since the header value should be optional

### Is there a workaround?

```
let auth: str? = req.headers?.get("authorization");
```

Explicitly marking it as optional string

### Component

SDK

### Wing Version

0.21.5

### Wing Console Version

_No response_

### Node.js Version

v18.16.0

### Platform(s)

MacOS

### Anything else?

https://winglang.slack.com/archives/C048QCN2XLJ/p1686943041584429

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #3141: optional chaining should create an optional type
### I tried this:

image

```ts
let j =  Json { };
let v = j.tryGet("working")?.tryGet("b")?.tryGet("c")?.asStr();
if v? {
//  |    ^ Expected optional type, found "str"
}
```

### This happened:

Failed to compile.

error: Expected optional type, found "str"
  --> test.w:3:4
  |
3 | if v? {
  |    ^ Expected optional type, found "str"

### I expected this:

v should be str? 

### Is there a workaround?

_No response_

### Component

Compiler

### Wing Version

0.22.35

### Wing Console Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by MarkMcCulloh**
Fixes #3141
Fixes #2965

Given `a.b` is a function with return type T:

- `a?.b`
  - Always an optional function
  - If assigned to a variable, can't be called (similar to TS)
- `a?.b()`:
  - Always `optional<T>`
  - Does not attempt to call b if a?.b is nil

To avoid any runtime failures, made sure to jsify more `?.`s

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributing/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by MarkMcCulloh** [comment]
This approach is actually unsafe, closing for now

File: examples/tests/valid/json.w [hunk 1]
```diff
@@ -146,7 +146,7 @@ let jsonElements = Json {
   }
 };
 
-if let val = jsonElements.tryGet("strings")?.tryGet("single")?.tryAsStr() {
+if let val = jsonElements.tryGet("strings")?.tryGet("single")?.asStr() {
```
Comment (staycoolcall911) [inline examples/tests/valid/json.w:149]: Could you please explain why you made that change?

Comment (MarkMcCulloh) [reply examples/tests/valid/json.w:149]: tryAsStr() returns `str?` while asStr() returns `str`. The change proves that the optionality of the previous calls to tryGet() carries forward into asStr(), which makes the file return value `str?`

Comment (staycoolcall911) [reply examples/tests/valid/json.w:149]: Got you, thanks!

**Review by staycoolcall911 — APPROVED** [review]
LGTM. I added a `pr/do-not-merge` label - to give @yoav-steinberg a chance to give his review.

**Review by yoav-steinberg — APPROVED** [review]

File: examples/tests/invalid/optionals.w [hunk 2]
```diff
@@ -83,3 +83,7 @@ let baz: Baz = Baz {
 // Ensure that ?. returns T? rather than T
 let val: str = baz?.bar?.foo?.val;
 //             ^^^^^^^^^^^^^^^^^^ Expected type to be "str", but got "str?" instead
+
+let optionalFunction = Json.tryParse("")?.asStr;
+optionalFunction();
+//^ Cannot call optional function (unless it's part of a reference)
```
[the remark this answers is not in the document]
Comment (MarkMcCulloh) [reply examples/tests/invalid/optionals.w:89]: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAWwIYE8BGBTAFASkQG8AoRcxGYRXAQigCcRtDSL3EHsoQGk2Og4GABcNQgF4AfMQC+ZQYnntZibABsAztmILBXHn0TgAJtmAww2E3qUl5JCAk1REUABaWA5ogkoMOAQA3I7OrpioXL5unmBeAPwAdMIh7AD0aQB68SQkGYgAKp6alCWo6upwAO7WJB7eScLBuflFMCXtiGBwruWVNTYRXMFAA

**Pull request merged**